Add Document about Informers and Listers (#2579)
googleforgames/agones: site/content/en/docs/Guides/access-api.md

	podInformer := informerFactory.Core().V1().Pods()

	// Create GameServer informer by informerFactory
	gameServers := agonesInformerFactory.Agones().V1().GameServers()
	gsInformer := gameServers.Informer()

	// Add EventHandler to informer
	// When the object's event happens, the function will be called
	// For example, when the pod is added, 'AddFunc' will be called and put out the "Pod Added"
	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
		AddFunc:    func(new interface{}) { logger.Infof("Pod Added") },
		UpdateFunc: func(old, new interface{}) { logger.Infof("Pod Updated") },
		DeleteFunc: func(old interface{}) { logger.Infof("Pod Deleted") },
	})
	gsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
		AddFunc:    func(new interface{}) { logger.Infof("GameServer Added") },
		UpdateFunc: func(old, new interface{}) { logger.Infof("GameServer Updated") },
		DeleteFunc: func(old interface{}) { logger.Infof("GameServer Deleted") },
	})

	// Start Go routines for informer
	informerFactory.Start(wait.NeverStop)
	agonesInformerFactory.Start(wait.NeverStop)
	// Wait until finish caching with List API
	informerFactory.WaitForCacheSync(wait.NeverStop)

> 	agonesInformerFactory.WaitForCacheSync(wait.NeverStop)

Reviewer: 🤔 do we want to use wait.NeverStop here, or do we want to drive people towards what they probably should do - which is have a `context.Context` (maybe cancellable? Might not matter, could just use [context.Background](https://pkg.go.dev/context#Background)), and pass in ctx.Done() into each of these?

Since it's likely better practice to use a `Context`. WDYT?

Also /cc @roberthbailey in case he has opinions.

Author: changed using `context.Background`.

```
You have just created a GameServer using Kubernetes Go Client.

> ## What are Informers and Listers? How can I use these?

Reviewer: ```suggestion
## Best Practice: Using Informers and Listers
```

> Informers and Listers are mechanism for Kubernetes.
> Almost all of the native Kubernetes controllers and custom controllers use these mechanisms.
> Kubernetes's control plane can be accessed via Kubernetes APIs, however,
> the load for control plane cannot be ignored when the client program accesses to the control plane every time.

Reviewer: ```suggestion
Almost all  Kubernetes' controllers and custom controllers utilise `Informers` and `Listers` to reduce the load on the.
Kubernetes's control plane.

Repetitive, direct access of the Kubernetes control plane API can significantly reduce the performance of the cluster -- and Informers and Listers help resolving that issue.
```

> Informers and Listers can reduce the load for the control plane by using the in-memory-cache.
> When Informes and Listers are created, the in-memory-cache called as the store are also created.
> The client program accesses and queries to the in-memory-cache, not control plane,
> which enables reducing the load.

Reviewer: ```suggestion
Informers and Listers reduce the load on the Kubernetes control plane by creating, using and maintaining and eventually consistent an in-memory-cache. This can be watched and also queried with zero cost, since it will only read against its in-memory model of the Kubernetes resources.
```

Informer's role and Lister's role are different.

> The Informer is the mechanism for watching Kubernetes object's event.
> When the Kubernetes objects changes(e.g. CREATE,UPDATE,DELETE), the `Event` object will create.
> The Informer lists and watches the object's changes through `Event` from the control plane.
> When the `Event` happens, the informer will watch it and store objects to the in-memory-cache.

Reviewer: ```suggestion

The Informer is the mechanism for watching a Kubernetes object's event.

When the Kubernetes objects changes(e.g. CREATE,UPDATE,DELETE), the Informer is informed, and can execute a  callback with the relevant object as an argument.

This can be very useful for building event based systems against the Kubernetes API.
```

Changed the wording here, since it's not an Event object - it's the actual resource object that comes through.

> The Lister is the mechanism for listing Kubernetes object's.
> Users can list the Kubernetes objects by using Kubernetes APIs, however,
> if users use List API, it causes the load for the control plane.
> The Informer stores the cache of the object to the in-memory-cache.
> If users use the in-memory-cache, the load can be avoidable.
> The Lister provides the list method while using the in-memory-cache.

Reviewer: ```suggestion
The Lister is the mechanism for querying Kubernetes object's against the client side in-memory cache.
Users can list the Kubernetes objects by using Kubernetes APIs, however,
if users do so, it places load on the Kubernetes control plane.

Since an Lister stores objects in an in-memory cache, queries against a come at practically no cost.
```

```
You have just created a GameServer using Kubernetes Go Client.

## Best Practice: Using Informers and Listers

Almost all  Kubernetes' controllers and custom controllers utilise `Informers` and `Listers`
to reduce the load on the Kubernetes's control plane.

Repetitive, direct access of the Kubernetes control plane API can significantly
reduce the performance of the cluster -- and Informers and Listers help resolving that issue.

Informers and Listers reduce the load on the Kubernetes control plane

> by creating, using and maintaining and eventually consistent an in-memory-cache.

Reviewer: ```suggestion
by creating, using and maintaining an eventually consistent an in-memory cache.
```

This can be watched and also queried with zero cost, since it will only read against
its in-memory model of the Kubernetes resources.

Informer's role and Lister's role are different.

> The Informer is the mechanism for watching a Kubernetes object's event.
> 
> When the Kubernetes objects changes(e.g. CREATE,UPDATE,DELETE), the Informer is informed,
> and can execute a callback with the relevant object as an argument.

Reviewer: ```suggestion

An Informer is the mechanism for watching a Kubernetes object's event, such that when an Kubernetes object changes(e.g. CREATE,UPDATE,DELETE), the Informer is informed,
and can execute a callback with the relevant object as an argument.
```

This can be very useful for building event based systems against the Kubernetes API.

> The Lister is the mechanism for querying Kubernetes object's against the client side in-memory cache.
> Users can list the Kubernetes objects by using Kubernetes APIs, however,
> if users do so, it places load on the Kubernetes control plane.
> 
> Since an Lister stores objects in an in-memory cache, queries against a come at practically no cost.

Reviewer: ```suggestion
Listers are a mechanism for querying Kubernetes object's against the client side in-memory cache.
Since an Lister stores objects in an in-memory cache, queries against a come at practically no cost.
```

I don't think we need to repeat the issue with direct acess.

Author: Changed from `Listers are a mechanism` to `A Lister is the mechanism` according to `An Informer is the mechanism`.
